src/x/serialize/unchecked_decoder.go

```diff
@@ -0,0 +1,156 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package serialize
+
+import (
+	"fmt"
+)
+
+// uncheckedDecoder is a fast decoder that skips all safety checks (i.e. ref counts).
+// it is suitable to use this when you are confident the provided bytes to decode are not shared.
+type uncheckedDecoder struct {
+	// immutable state for the lifetime of the decoder
+	maxTags      uint16
+	maxTagLength uint16
+
+	// immutable state for the lifetime of a single use.
+	length int
+
+	// mutable state during a single use.
+	data            []byte
+	remaining       int
+	err             error
+	currentTagName  []byte
+	currentTagValue []byte
+}
+
+func newUncheckedTagDecoder(tagLimits TagSerializationLimits) *uncheckedDecoder {
+	return &uncheckedDecoder{
+		// inline the config values since this call is in the hotpath of the matcher. saved about 1% cpu.
+		maxTags:      tagLimits.MaxNumberTags(),
+		maxTagLength: tagLimits.MaxTagLiteralLength(),
+	}
+}
+
+func (d *uncheckedDecoder) reset(b []byte) {
+	d.data = b
+	var length uint16
+	if len(d.data) == 0 {
+		d.length = 0
+		d.remaining = 0
+	} else {
+		header, err := d.decodeUInt16()
+		if err != nil {
+			d.err = err
+			return
+		}
+
+		if header != HeaderMagicNumber {
+			d.err = ErrIncorrectHeader
+			return
+		}
+
+		length, err = d.decodeUInt16()
+		if err != nil {
+			d.err = err
+			return
+		}
+
+		if limit := d.maxTags; length > limit {
+			d.err = fmt.Errorf("too many tags [ limit = %d, observed = %d ]", limit, length)
+			return
+		}
+	}
+	d.length = int(length)
+	d.remaining = int(length)
+	d.err = nil
+	d.currentTagName = nil
+	d.currentTagValue = nil
+}
+
+func (d *uncheckedDecoder) next() bool {
+	if d.err != nil || d.remaining <= 0 {
+		return false
+	}
+
+	if err := d.decodeTag(); err != nil {
+		d.err = err
+		return false
+	}
+
+	d.remaining--
+	return true
+}
+
+func (d *uncheckedDecoder) current() ([]byte, []byte) {
+	return d.currentTagName, d.currentTagValue
+}
+
+func (d *uncheckedDecoder) decodeTag() error {
+	var err error
+	d.currentTagName, err = d.decodeID()
+	if err != nil {
+		return err
+	}
+	if len(d.currentTagName) == 0 {
+		return ErrEmptyTagNameLiteral
+	}
+
+	d.currentTagValue, err = d.decodeID()
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *uncheckedDecoder) decodeID() ([]byte, error) {
+	l, err := d.decodeUInt16()
+	if err != nil {
+		return nil, err
+	}
+
+	if limit := d.maxTagLength; l > limit {
+		return nil, fmt.Errorf("tag literal too long [ limit = %d, observed = %d ]", limit, int(l))
+	}
+
+	if len(d.data) < int(l) {
+		return nil, errInvalidByteStreamIDDecoding
```

Comment:
Is it valid to have a tag with zero value length? If so this logic breaks. 

I think technically we allow that but we probably shouldn’t, although something perhaps to check against the other decoder.

Reply:
Ah, looks like this is the same 👍 
https://github.com/m3db/m3/blob/2b93a5e159a9f2005b2f904d777ff7123637edda/src/x/serialize/decoder.go#L155-L157

Should retain same logic then 👍

Reply:
lol yea I copied that original decoder to start :)